plugins/experimental/remap_purge/remap_purge.c

```diff
@@ -0,0 +1,335 @@
+    TSHandleMLocRelease(reqp, TS_NULL_MLOC, hdr_loc);
+  }
+
+  /* Setup the continuation to handle this request if appropriate, if not, set the GenID if needed */
+  if (should_purge) {
+    TSCont cont = TSContCreate(purge_cont, TSMutexCreate());
+
+    TSContDataSet(cont, purge);
+    TSHttpTxnHookAdd(txnp, TS_HTTP_CACHE_LOOKUP_COMPLETE_HOOK, cont);
+    TSHttpTxnHookAdd(txnp, TS_HTTP_SEND_RESPONSE_HDR_HOOK, cont);
+  } else if (purge->gen_id > 0) {
+    TSHttpTxnConfigIntSet(txnp, TS_CONFIG_HTTP_CACHE_GENERATION, purge->gen_id);
+  }
+}
+
+TSReturnCode
+TSRemapInit(TSRemapInterface *api_info, char *errbuf, int errbuf_size)
+{
+  TSDebug(PLUGIN_NAME, "initialized");
+  return TS_SUCCESS;
+}
+
+TSReturnCode
+TSRemapNewInstance(int argc, char *argv[], void **ih, char *errbuf, int errbuf_size)
+{
+  char *id                             = argv[0]; /* The ID is default to the "from" URL, so save it */
+  PurgeInstance *purge                 = TSmalloc(sizeof(PurgeInstance));
+  static const struct option longopt[] = {
+    {(char *)"id", required_argument, NULL, 'i'},     {(char *)"secret", required_argument, NULL, 's'},
+    {(char *)"header", required_argument, NULL, 'h'}, {(char *)"state_file", required_argument, NULL, 'f'},
+    {(char *)"allow_get", no_argument, NULL, 'a'},    {NULL, no_argument, NULL, '\0'},
+  };
+
+  memset(purge, 0, sizeof(PurgeInstance));
+
+  // The first two arguments are the "from" and "to" URL string. We need to
+  // skip them, but we also require that there be an option to masquerade as
+  // argv[0], so we increment the argument indexes by 1 rather than by 2.
+  argc--;
+  argv++;
+  optind = 0;
+
+  for (;;) {
+    int opt = getopt_long(argc, (char *const *)argv, "", longopt, NULL);
+
+    if (opt == -1) {
+      break;
+    }
+
+    switch (opt) {
+    case 'a':
+      purge->allow_get = true;
+      break;
+    case 'h':
+      purge->header     = TSstrdup(optarg);
+      purge->header_len = strlen(purge->header);
+      break;
+    case 'i':
+      purge->id = TSstrdup(optarg);
+      break;
+    case 's':
+      purge->secret     = TSstrdup(optarg);
+      purge->secret_len = strlen(purge->secret);
+      break;
+    case 'f':
+      purge->state_file = make_state_path(optarg);
+      break;
+    }
+  }
+
+  if ((NULL == purge->secret) || (NULL == purge->state_file) || !purge->secret_len) {
+    TSError("[%s] Unable to create remap instance, need at least a secret (--secret) and state (--state_file)", PLUGIN_NAME);
+    return TS_ERROR;
+  } else {
+    init_purge_instance(purge, id);
+    *ih = (void *)purge;
+    return TS_SUCCESS;
+  }
+}
+
+void
+TSRemapDeleteInstance(void *ih)
+{
+  PurgeInstance *purge = (PurgeInstance *)ih;
+
+  delete_purge_instance(purge);
+}
+
+TSRemapStatus
+TSRemapDoRemap(void *ih, TSHttpTxn txnp, TSRemapRequestInfo *rri)
+{
+  PurgeInstance *purge = (PurgeInstance *)ih;
+
+  if (purge) {
+    handle_purge(txnp, purge);
+  } else {
+    TSReleaseAssert(!"No instance data");
+  }
```

Comment:
Suggestion; since you can't get ``NULL`` here, just simplify by removing the checks.